---
title: "Supplementary data"
output:
  html_document:
    theme: cerulean
---

```{r setup, echo=FALSE, include=FALSE}
library(knitr)
library(DESeq2)
opts_knit$set(root.dir = "~/GRewd/pipeline")
```

```{r loadData, echo=FALSE}
####
# load data
#

source("~/GRewd/pipeline/R/orthoGrpTools.R")

orthoPath <- "/mnt/NOBACKUP/mariansc/share/orthos"
grps <- loadOrthoGrpsArray(file.path(orthoPath,"orthoMCL/groups.txt"))
goodGrps <- loadOrthoGrpsArray(file.path(orthoPath,"splitGroups/goodGroups.txt"))

exprHCOG <- readRDS("~/GRewd/pipeline/data/superExprMatrix.RDS") # from: superGeneModel.Rmd
DEmat <- readRDS("~/GRewd/pipeline/data/DEmat.RDS")  # from: superGeneModel.Rmd

load(file.path(orthoPath,"DESeq/DE.RData"))
```


```{r defOutputFiles, echo=FALSE}
####
# output files
#

allOrthoGroupsFile <- "SupplementaryData/allOrthoGroups.tsv"
highConfOrthoGroupsFile <- "SupplementaryData/HighConfOrthoGroups.tsv"
HCOGReadCountsFile <- "SupplementaryData/HCOGreadCounts.tsv"
HCOG_DEresultsFile <- "SupplementaryData/HCOG_DEresults.tsv"
```


### allOrthoGroups.tsv

Ortholog groups generated by OrthoMCL are stored as a table with orthogroup IDs in rows and species in columns. Each cell contains sequence IDs separated by ",". Column description:


|Species ID  |Description                                   |
|:-----------|:---------------------------------------------|
|HoVu        |_H. vulgare_ (denovo)                         |
|BrDi        |_D. distachyon_ (denovo)                      |
|MeNu1       |_M. nutens_ (denovo)                          |
|StLa        |_S. lagascae_ (denovo)                        |
|NaSt        |_N. stricta_ (denovo)                         |
|Hv_R        |_H. vulgare_ (reference)                      |
|Bd_R        |_D. distachyon_ (reference)                   |
|Os_R        |_O. sativa_ (reference)                       |
|Zm_R        |_Z. mays_ (reference)                         |
|Sb_R        |_S. bicolor_ (reference)                      |
|LoPe        |_L. perenne_ (reference denovo)               |
|LoPeF       |_L. perenne_ var. falster (reference denovo)  |


```{r allOGs, echo=FALSE}
####
#
# First for all orthogroups
#

# reorder the species into a slightly more logical order
spcOrder <- c("HoVu","BrDi","MeNu1","StLa","NaSt","Hv_R","Bd_R","Os_R","Zm_R","Sb_R","LoPe","LoPeF")

if(!file.exists(allOrthoGroupsFile)){
  grpsCSP <- sapply(grps,paste,collapse=",") # make comma separated paralogs
  dim(grpsCSP) <- dim(grps)
  dimnames(grpsCSP) <- dimnames(grps)
  
  write.table(cbind(grpID=rownames(grpsCSP),grpsCSP[,spcOrder]),file = allOrthoGroupsFile, 
              sep = "\t",quote=F,row.names = F,)
}

```

### HighConfOrthoGroups.tsv

High confidence ortholog groups (HCOGs) generated by filtering and splitting orthogroups. Same format as described above. Sub-groups resulting from splitting larger groups are designated with a numbered suffix in the group ID.


```{r HCOGs, echo=FALSE}
####
#
# HCOGs
#

if(!file.exists(highConfOrthoGroupsFile)){
  
  # Extract the HCOGs from the goodGrps
  HCgrps <- goodGrps[rownames(exprHCOG),]
  
  HCgrpsCSP <- sapply(HCgrps,paste,collapse=",") # make comma separated paralogs
  dim(HCgrpsCSP) <- dim(HCgrps)
  dimnames(HCgrpsCSP) <- dimnames(HCgrps)
  
  write.table(cbind(grpID=rownames(HCgrpsCSP),HCgrpsCSP[,spcOrder]),file = highConfOrthoGroupsFile, 
              sep = "\t",quote=F,row.names = F,)
}

```

### HCOGreadCounts.tsv

Combinened readcounts for high condfidence ortholog groups. Each column represents a sample, and rows represents HCOGs. The sample IDs in the column header consists of the species ID, the timepoint, whether the sample is pooled from five individual plants ("mix") or just a single individual plant ("ind") and a replicate number.


```{r HCOG_readcounts, echo=FALSE}
####
#
# HCOG summed paralogs read counts
#

# rename samples
renameSamples <- function(sampleNames){
  for( substitusion in list( c("NaSt1","NaSt"),
                             c("T-1","W0"),
                             c("T0","D0"),
                             c("T1","D1"),
                             c("T3","W4"),
                             c("T4","W9"))){
    sampleNames <- sub(substitusion[1],substitusion[2],sampleNames)
  }
  return(sampleNames)
}
 
colnames(exprHCOG) <- renameSamples(colnames(exprHCOG))

write.table(cbind(grpID=rownames(exprHCOG),exprHCOG),file = HCOGReadCountsFile, 
            sep = "\t",quote=F,row.names = F,)

```

### HCOG_DEresults.tsv

Differential expression test results for HCOGs. Column description:

* ST - short-term response
* LT - long-term response
* lfc - log2 fold change
* pVal - P-value
* pAdj - Benjamini-Hochberg adjusted P-value

```{r HCOG_DEresults, echo=FALSE}
####
#
# HCOG DE results
#

# Everything into one table
# columns: [effect].[lfc, pVal, pAdj].[spc] 

# rename peak=ST (short term response), ramp=LT (long term response)

# function to cbind a list of tables while prefixing the names from the list to the colnames
cTables <- function(listOfTables,sep="."){
  retVal <- Reduce(cbind,listOfTables)
  colnames(retVal) <- paste(rep(names(listOfTables),sapply(listOfTables,ncol)),colnames(retVal),sep=sep)
  return(retVal)
}

HCOG_DE <-  cTables( list( ST = cTables(listOfTables = DEmat$peak[c("lfc", "pVal", "pAdj")]),
                           LT = cTables(listOfTables = DEmat$ramp[c("lfc", "pVal", "pAdj")])))

write.table(cbind(grpID=rownames(HCOG_DE),HCOG_DE),file = HCOG_DEresultsFile, 
            sep = "\t",quote=F,row.names = F,)

```

### *_readCounts.tsv

Readcounts for all genes included in ortholog groups. One table for each species. SampleIDs in column headers corresponds to the ones in HCOGreadCounts.tsv.

```{r Spc_readCounts, echo=FALSE}
####
#
#  All orthogroups read counts
#

# Only transcripts in orthogroups
readCountFiles <- dir(file.path(orthoPath,"exprTbls"),pattern="^[^_]*_expected_countTbl.txt",full.names = T)

for(inFile in readCountFiles){
  outFile <- sub("_expected_countTbl.txt","_readCounts.tsv",basename(inFile))
  x <- readLines(inFile)
  sampleNames <- renameSamples(unlist(strsplit(x[1],split="\t")))
  x[1] <- paste(c("seqID",sampleNames),collapse="\t")
  writeLines(x,con=file.path("SupplementaryData",outFile))
}

```

### *_readCounts.tsv

Differential expression results for all genes included in ortholog groups.  One table for each species. Column header format similar to that in HCOG_DEresults.tsv.

```{r Spc_DEresults, echo=FALSE}
####
#
# Species specific DE results
#

# columns: [ST,LT].[lfc, pVal, pAdj]

for(spc in names(DE)){
  x <- DE[[spc]][c("resPeak","resRamp")]
  names(x) <- c("ST","LT")
  x <- lapply(x, function(y){
    y <- as.matrix(y[ ,c("log2FoldChange","pvalue","padj")])
    colnames(y) <- c("lfc", "pVal", "pAdj")
    return(y)
  })
  x <- cTables(x)
  write.table(cbind(grpID=rownames(x),x),file = file.path("SupplementaryData",paste0(spc,"_DEresults.tsv")), 
              sep = "\t",quote=F,row.names = F,)
}

```

### denovo transcriptomes.zip

Contains fasta files with the transcriptomes generated by Trinity for each of the five species.