---
title: "GO analysis using topGO"
output:
  html_document:
    toc: yes
---

```{r loadLibraries, echo=F, message=F}
# # un-comment to install from bioconductor:
# source("http://bioconductor.org/biocLite.R")
# biocLite("topGO")
library(topGO)
library(DESeq2)
library(gplots)


source("~/GRewd/pipeline/R/orthoGrpTools.R")
```


```{r loadGOannot, echo=FALSE}
GOannotFile <- "/mnt/NOBACKUP/mariansc/share/refGenomes/barley_HighConf_genes_MIPS_23Mar12_IPROScan_GOs.txt"
GOannotFileSource <- "ftp://ftpmips.helmholtz-muenchen.de/plants/barley/public_data/genes/barley_HighConf_genes_MIPS_23Mar12_IPROScan_GOs.txt"

if(!file.exists(GOannotFile)){
  download.file(GOannotFileSource, destfile=GOannotFile)
}

# read GO annotations
HvGOtbl <- read.table(GOannotFile,sep="\t", quote = "",
                      col.names=c("geneID","GO"), stringsAsFactors=F)
# convert to gene2GO list
HvGene2GO <- split(HvGOtbl$GO,HvGOtbl$geneID)

```

```{r loadData, echo=FALSE}
# load DE results
load("/mnt/NOBACKUP/mariansc/share/orthos/DESeq/DE.RData")

grps <- loadOrthoGrpsArray("/mnt/NOBACKUP/mariansc/share/orthos/orthoMCL/groups.txt")

lens <- sapply(grps,length)
dim(lens) <- dim(grps)
dimnames(lens) <- dimnames(grps)
```


```{r defGOFunctions, echo=FALSE}

geneIDs2grpIDs <- function(geneIDs, spc, grps){
  flatGrpIDs <- rep(rownames(grps),sapply(grps[ ,spc],length))
  spcGeneIDs <- unlist(grps[,spc],use.names = F)
      
  flatGrpIDs[match(geneIDs,spcGeneIDs)]
}

# make grp2GO list
HvGOtbl$grpID <- geneIDs2grpIDs(HvGOtbl$geneID,"Hv_R",grps)
grpID2GO <- lapply( split( HvGOtbl$GO, HvGOtbl$grpID ), unique )


testGOgrps <- function(grpIDs, bkGrpIDs = NULL, ontology = "BP"){
  # get annotated background groups
  if( !is.null(bkGrpIDs) ){
    annBkGrpIDs <- intersect(names(grpID2GO),bkGrpIDs)
  } else {
    annBkGrpIDs <- names(grpID2GO)
  }
  
  allGenes <- factor(as.integer(annBkGrpIDs %in% grpIDs))
  names(allGenes) <- annBkGrpIDs
  
  sink("/dev/null") # silence the output from topGO
  
  GOdata <- new("topGOdata", ontology = ontology, allGenes = allGenes,
                annot = annFUN.gene2GO, gene2GO = grpID2GO)
  
  
  # run fisher test using classic algorithm
  classic.fisher <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
  weight.fisher <- runTest(GOdata, algorithm = "weight", statistic = "fisher")
  
  sink() # stop sinking output
  
  GenTable(GOdata, classic = classic.fisher, weight=weight.fisher,
           orderBy = "weight", ranksOf="classic", topNodes = 20)
}
```

```{r plotVenn, echo=FALSE}
# get grpIDs of significant genes
lapply(c(ramp="resRamp",peak="resPeak"),function(eff){
  lapply(c(up=1,down=-1),function(direction){
    lapply(setNames(names(DE),names(DE)), function(spc){
      with(DE[[spc]][[eff]], {
        (padj<0.05) * sign(log2FoldChange) * (abs(log2FoldChange)>1)
      }) -> sig
      sig[is.na(sig)] <- 0
      
      flatGrpIDs <- rep(rownames(grps),lens[ ,spc])
      spcGeneIDs <- unlist(grps[,spc])
      
      sigGeneIDs <- rownames(DE[[spc]][[eff]])[sig==direction]
      sigGrpIDs <- flatGrpIDs[match(sigGeneIDs,spcGeneIDs)]
      
      return( unique(sigGrpIDs))
    })
  }) 
}) -> sigGrpIds

```

### GO enrichment of common response orthologs

Let's look at the genes that are up regulated in all species (center of venn diagram):
```{r upPeakInAllVennGO, echo=F, meassage=F}
venn( data=sigGrpIds$peak$up[c("BrDi","HoVu","NaSt","MeNu1","StLa")])

title( main="Significant peak effect",
       sub="adj.P < 0.05, log2FC > 1")

grpIDs <- Reduce(intersect,sigGrpIds$peak$up[c("BrDi","HoVu","NaSt","MeNu1","StLa")])

DT::datatable(testGOgrps(grpIDs))
```


Let's look at the ramp up effect
```{r upRampInAllVennGO, echo=F, meassage=F}
venn( data=sigGrpIds$ramp$up[c("BrDi","HoVu","NaSt","MeNu1","StLa")])

title( main="Significant ramp effect",
       sub="adj.P < 0.05, log2FC > 1")

grpIDs <- Reduce(intersect,sigGrpIds$ramp$up[c("BrDi","HoVu","NaSt","MeNu1","StLa")])
# grpIDs <- Reduce(intersect,sigGrpIds$ramp$up[c("BrDi","HoVu","NaSt","MeNu1")])
# grpIDs <- Reduce(intersect,sigGrpIds$ramp$up[c("BrDi","HoVu")])

DT::datatable(testGOgrps(grpIDs))
```

Lets test the GO enrichment in groups that have at least one ortholog in each of the species
```{r echo=FALSE, message=FALSE}
grpIDsInAll <- rownames(lens)[apply(lens[ , c( "BrDi", "HoVu", "NaSt", "MeNu1", "StLa")]>0,1,all)]
DT::datatable(testGOgrps(grpIDsInAll))
```

Lets re-run the ramp up GO enrichment using only the orthologs existing in all as background.
```{r echo=FALSE, message=FALSE}

DT::datatable(testGOgrps(grpIDs,bkGrpIDs = grpIDsInAll))
```
