---
title: "readCountOverview"
output:
  html_document:
    toc: yes
---

# New samples
```{r loadLibraries, echo =FALSE}
library(RColorBrewer)
library(stringr)
library(ggplot2)
source("/mnt/users/lagr/GRewd/pipeline/R/orthoGrpTools.R")
```

```{r loadData, echo=FALSE}
pipelineOutDir <- "/mnt/NOBACKUP/mariansc/share"
grps <- loadOrthoGrpsArray(orthoGrpFile = file.path(pipelineOutDir,"orthos/orthoMCL/groups.txt"))

tblDir <- file.path(pipelineOutDir,"orthos/exprTbls")

grp11Cnt <- read.table(file = file.path(tblDir,"grpSingletonCountTbl.txt"), 
                       stringsAsFactors = F)
# grp11FPKM <- read.table(file = file.path(tblDir,"grpSingletonFPKMTbl.txt"), 
#                        stringsAsFactors = F)
grpSumCnt <- read.table(file = file.path(tblDir,"grpSumCountTbl.txt"), 
                       stringsAsFactors = F)
# grpSumFPKM <- read.table(file = file.path(tblDir,"grpSumFPKMTbl.txt"), 
#                        stringsAsFactors = F)

files <- dir(tblDir,pattern="_full_expected_countTbl.txt",full.names = T)
exprCntFull <- lapply(setNames(files,str_extract(basename(files),"^[^_]+")), 
                  read.table, stringsAsFactors = F)

files <- dir(tblDir,pattern="_expected_countTbl.txt",full.names = T)
files <- files[!grepl("_full_",files)]
exprCnt <- lapply(setNames(files,str_extract(basename(files),"^[^_]+")), 
                  read.table, stringsAsFactors = F)


# files <- dir(tblDir,pattern="_FPKMTbl.txt",full.names = T)
# exprFPKM <- lapply(setNames(files,str_extract(basename(files),"^[^_]+")), 
#                   read.table, stringsAsFactors = F)

spcs <- names(exprCnt)
spcs <- setNames(spcs,spcs)

```

### Read count distributions

#### Ortholog groups

The reads have been divided into those that have exactly one ortholog per species; those that have one or more per species; those that have at least one ortholog in any species.

```{r histoPlots, echo=FALSE}
lens <- unlist(lapply(grps,length))
dim(lens) <- dim(grps)
dimnames(lens) <- dimnames(grps)
grpsInAll <- apply(lens>0,1,all)
grps111 <- apply(lens==1,1,all)
grpsInDeNovo <- apply(lens[,spcs]>0,1,all)
grps111DeNovo <- apply(lens[,spcs]==1,1,all)




for(spc in spcs){
  x <- as.matrix(exprCntFull[[spc]])
  x <- ifelse(x<1,0,x)
  x <- log2(x)
  
  hist(x,xlab="log2(count)",main=spc,breaks=seq(0,20,by=1),col="lightblue")

  isInGrps <- rownames(x) %in% unlist(grps[,spc])

  lines(hist(x[isInGrps,],plot=FALSE,breaks=seq(0,20,by=1)),col="rosybrown2")
  
  isInAllGrps <- rownames(x) %in% unlist(grps[grpsInDeNovo,spc])
  
  lines(hist(x[isInAllGrps,],plot=FALSE,breaks=seq(0,20,by=1)),col="purple")

  is111 <- rownames(x) %in% unlist(grps[grps111DeNovo,spc])
  
  lines(hist(x[is111,],plot=FALSE,breaks=seq(0,20,by=1)),col="red")

  legend("topright",legend=c("All transcripts","In orthogroup","In all de-novo species","Single in all de-novo"),
         fill=c("lightblue","rosybrown2","purple","red"))

}
```

#### Per sample
(Only genes that belong to group)

```{r singleSampleDensity, echo=FALSE,fig.height=9,fig.width=9}

for(spc in spcs){

  apply(exprCnt[[spc]],2, function(x){
    x <- log2(x)
    return(density(x,))
  }) -> dens
  
  plot(x=NULL, xlim=c(0,20),ylim=c(0,0.15), xlab="log2(count)",ylab="Density",main=spc)
  cols=rainbow(length(dens))

  for( i in 1:length(dens)){
    lines(dens[[i]],col=cols[i])
  }
  
  legend("topright",legend=colnames(exprCnt[[spc]]), lty=lty,
         lwd=2,col=cols[1:length(dens)])

}


```

### Clustering of samples

#### Samples within species

Filtering: only use genes with mean count higher than 10

```{r withinSpeciesClustering, echo=FALSE}
# cluster within each assembly

for(spc in spcs){

  # make table of counts
  countTbl <- as.matrix(exprCnt[[spc]])


  countTbl.filt <- countTbl[ rowMeans(countTbl) > 10,]
#   countTbl.sqrt <- sqrt(countTbl.filt)
  countTbl.sqrt <- log2(1+countTbl.filt)

  # calculate correlation between samples
  corr <- cor(countTbl.sqrt)

    # plot dendrogram
  plot(hclust(as.dist(1-corr)),main=spc)
  heatmap(corr,scale = "none")
  
}
```

#### Across species samples clustering

```{r acrossSpeciesClustering, echo=FALSE, fig.height=8, fig.width=12}


# spcColors <- c(BrDi="#00A0B0", HoVu="#6A4A3C", MeNu1="#CC333F",
#                MeNu2="#EB6841", NaSt="#EDC951", StLa="#2ECC71")
spcColors <- c(BrDi="#00A0B0", HoVu="#6A4A3C", MeNu1="#CC333F",
               MeNu2="#EB9891", NaSt="#EDC951", StLa="#2ECC71")


# dim(x)
# x <- as.matrix(grp11Cnt)
# # x <- x[ rowMeans(x) > 10,]
# x <- sqrt(x)
# pearsonCorr <- cor(x)
# 
# hc <- hclust(as.dist(1-pearsonCorr))
# plot(hc,hang=-1,main="Clustering, 1:1 ortholog groups, sqrt(read counts)")
# points(x=1:length(hc$labels),y=rep(0,length(hc$labels)),
#        pch=20,cex=3,
#        col=spcColors[str_extract(hc$labels[hc$order],paste(spcs,collapse="|"))])
# points(x=1:length(hc$labels),y=rep(0,length(hc$labels)),
#        pch=c(T0="0",T1="1",T3="3",T4="4")[str_extract(hc$labels[hc$order],"T[0-4]")],
#        ,cex=1.1,col="white")

x <- as.matrix(grpSumCnt[grps111DeNovo,])
x <- x[ rowMeans(x) > 10,]
x <- log2(1+x)

pearsonCorr <- cor(x)
hc <- hclust(as.dist(1-pearsonCorr))
plot(hc,hang=-1,main="Clustering, 1:1 de-novo ortholog groups, log2(read counts)")
points(x=1:length(hc$labels),y=rep(0,length(hc$labels)),
       pch=20,cex=3,
       col=spcColors[str_extract(hc$labels[hc$order],paste(spcs,collapse="|"))])
points(x=1:length(hc$labels),y=rep(0,length(hc$labels)),
       pch=c(T0="0",T1="1",T3="3",T4="4")[str_extract(hc$labels[hc$order],"T[0-4]")],
       ,cex=1.1,col="white")

```

```{r acrossSpeciesHeatmap, echo=FALSE, fig.height=12, fig.width=12,eval=FALSE}
heatmap(pearsonCorr,scale = "none")
```

### PCA
Based on log2(1+counts) of 1:1 ortholog groups in all de-novo transcriptomes.

```{r pcaPlot, echo=FALSE}
#### HOW TO DO IT WITH VSD
#
# x <- as.matrix(grpSumCnt[grpsInAll,])
# 
# pDat = data.frame( spc = str_extract(colnames(x),paste(spcs,collapse="|")),
#                    timePoint = str_extract(colnames(x),"T[0-4]") )
# 
# dds <- DESeqDataSetFromMatrix(countData = round(x),
#                               colData = pDat,
#                               design = formula(~ 1))
# vsd <- varianceStabilizingTransformation(dds)
# x <- assay(vsd)

x <- as.matrix(grpSumCnt[grps111DeNovo,])
x <- x[ rowMeans(x) > 10,]
x <- log2(1+x)

pDat = data.frame( spc = str_extract(colnames(x),paste(spcs,collapse="|")),
                   timePoint = sub("\\.","-",str_extract(colnames(x),"T\\.?[0-4]")) )

pch=sub("-","\2",substr(pDat$timePoint,2,2))

pca <- prcomp(t(x))
plot(pca)

plot(pca$x[,"PC1"],pca$x[,"PC2"], col=spcColors[pDat$spc], pch=pch)

legend("topright",legend=levels(pDat$spc),fill=spcColors[levels(pDat$spc)])
legend("topleft",legend=unique(pDat$timePoint),pch=unique(pch))

plot(pca$x[,"PC3"],pca$x[,"PC4"], col=spcColors[pDat$spc], pch=pch)
plot(pca$x[,"PC5"],pca$x[,"PC6"], col=spcColors[pDat$spc], pch=pch)
plot(pca$x[,"PC7"],pca$x[,"PC8"], col=spcColors[pDat$spc], pch=pch)

```
