---
title: "Tree with candidate LTI gene groups"
output: html_document
---


```{r loadLibraries, echo=FALSE, message=FALSE}
library(DESeq2)
library(gplots)
library(stringr)
library(ape)


source("~/GRewd/pipeline/R/orthoGrpTools.R")
```


```{r loadData, echo=FALSE}
orthoPath <- "/mnt/NOBACKUP/mariansc/share/orthos"


dNdSall <- readRDS("~/GRewd/pipeline/data/dNdSall.RDS") # from pamlResultsOverview.Rmd
pamlRes <- readRDS("~/GRewd/pipeline/data/pamlRes.RDS") # from pamlResultsOverview.Rmd 

hTestable <- readRDS("~/GRewd/pipeline/data/hTestable2.RDS") # from treeAnalysis6.Rmd

# convert pamlResult into pvalue matrix with only testable hypotheses
pamlMat <- matrix(NA_real_,nrow = nrow(hTestable),ncol = ncol(hTestable), 
                  dimnames = dimnames(hTestable))

for(h in colnames(hTestable)){
  testableGrpIDs <- rownames(hTestable)[hTestable[ ,h]]
  pamlMat[testableGrpIDs,h] <- pamlRes[[h]]$pVal[match(testableGrpIDs, pamlRes[[h]]$grpID)]
}

# load DEmat based based on superGenes
DEmat <- readRDS("~/GRewd/pipeline/data/DEmat.RDS")  # from: superGeneModel.Rmd



# load groups
goodGrps <- loadOrthoGrpsArray(file.path(orthoPath,"splitGroups/goodGroups.txt"))

```


### Combined peak and ramp

A gene is considered cold responsive if it is significantly differencially expressed (P-adjusted < 0.05 and abs(log fold change) > 1) either in peak or ramp.

```{r sigVennPlet, echo=FALSE}
# Get significant genes for peak and ramp (NA values equals not significant)
sigMatUpDown <- lapply(DEmat, with, { ifelse(is.na(pAdj), F, pAdj < 0.05 & abs(lfc) > 1) })
# Combine peak and ramp (significant if significant in either peak or ramp)
combSigMat <- sigMatUpDown$peak | sigMatUpDown$ramp

# plot Venn diagram
venn(apply(combSigMat,2,which))
title(main="significant peak or ramp (up or down)",sub="p.adj<0.05 & |lfc|>1")
```


```{r lolium, echo=FALSE, eval=FALSE}
# ### Lolium data
# 
# Data from the article [Vernalization Mediated Changes in the Lolium perenne Transcriptome](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0107365)
# 
# We used the data from leafs of the Falster variant.
#  

# load variance stabilized lolium (Falster) expression
loliumVSD <- read.table("~/GRewd/pipeline/indata/LoliumVSD/Fal_leaf_DE_VSD.txt",header=T,sep="\t",row.names = 1)




# convert lolium sequence ID's
loliumVSDseqIDs <- sub("comp([0-9]+)_c([0-9]+)_seq([0-9]+)","f\\1_c\\2s\\3",rownames(loliumVSD),perl=T)

# how many have ortholog in good group?
table(loliumGeneInGoodOrthoGroup=loliumVSDseqIDs %in% unlist(goodGrps[ ,"LoPeF"]))

# get grpIDs
loliumVSDgrpIDs <- seqIDtoGrpID(goodGrps,loliumVSDseqIDs,"LoPeF")

# How many sequences per group (paralogs)?
table(paralogCount=table(na.omit(loliumVSDgrpIDs)))

# get unique grps
loliumSigGrpIDs <- unique(na.omit(loliumVSDgrpIDs))

# How many of these groups are included in the superGeneMatrix?
table(loliumGrpIdInSuperGeneMatrix = loliumSigGrpIDs %in% rownames(combSigMat))


# # create a matrix with only
# loliumVSDmat <- as.matrix(loliumVSD)
# rownames(loliumVSDmat) <- loliumVSDseqIDs
# loliumVSDmat <- loliumVSDmat[rownames(loliumVSDmat) %in% unlist(goodGrps[ ,"LoPeF"]),]
# loliumVSDmatGrpIDs <- seqIDtoGrpID(goodGrps,rownames(loliumVSDmat),"LoPeF")

loliumSig <- rownames(combSigMat) %in% loliumSigGrpIDs


# plot Venn diagram
venn(apply(combSigMat & !loliumSig,2,which))
title(main="significant but not in lolium")

venn(apply(combSigMat & loliumSig,2,which))
title(main="significant also in lolium",
      sub=paste("Only in lolium =",sum(loliumSig & !apply(combSigMat,1,any))))


```


```{r calcSig, echo=FALSE}
# get species specific LTI genes
spcs <- colnames(combSigMat)
sapply( spcs, function(spc){
  # significant in spc and not significant in any other spc 
  combSigMat[ ,spc] & !apply(combSigMat[ ,spcs != spc],1,any)
}) -> spcSig


# Species included in each branch/hypothesis (excluding lolium)
hSpcs2 <- list(
  H4b = c("HoVu","BrDi"),
  #H4e = c("HoVu","BrDi","MeNu1"),
  H4d = c("HoVu","BrDi","MeNu1","StLa"),
  H4a = c("HoVu","BrDi","MeNu1","StLa","NaSt")
)

# Get branch specific LTI genes
sapply( hSpcs2, function(branchSpcs){
  # significant in all branch species and not significant in any other species 
  apply(combSigMat[ ,spcs %in% branchSpcs],1,all) & 
    !apply(combSigMat[ ,!(spcs %in% branchSpcs),drop=F],1,any)
}) -> hSig

```

```{r treePlot, echo=FALSE}
# New species tree with collapsed Stipa/Melica
spcTree <- read.tree(text = "((((HoVu:1,BrDi:1)H4b:1,MeNu1:2,StLa:2)H4d:1,NaSt:3)H4a:1,Os_R:4):0;")

# get edge numbers for the branch specific edges
branchNodeNr <- match(colnames(hSig),spcTree$node.label) + length(spcTree$tip.label)
branchEdgeNr <- setNames( match(branchNodeNr,spcTree$edge[ ,2]), colnames(hSig) )

# Species names
spcLabels <- c( Os_R = "Oryza",
                NaSt = "Nardus",
                StLa = "Stipa",
                MeNu1 = "Melica",
                BrDi = "Brachypodium",
                HoVu = "Hordeum") #,LoPe = "Lolium")


svg(filename = "~/GRewd/pipeline/articleFigures/FigureGO_SpcTree.svg",
    width = 6, height = 3)

par(mar=c(1,1,1,1) + 0.1)
plot.phylo(spcTree,edge.width = 3,show.tip.label = F,x.lim=c(0,7),y.lim=c(0,7))

for( i in seq_along(spcLabels)){
  tiplabels( bquote(italic(.(spcLabels[i]))), tip = match(spcTree$tip.label, names(spcLabels))[i] ,
             adj = 0,frame = "none",cex=1.2)  
}
par(mar=c(5, 4, 4, 2) + 0.1)

invisible(dev.off())
  
# 
# tiplabels( paste(" ",spcLabels), tip = match(spcTree$tip.label, names(spcLabels)) ,
#            adj = 0,frame = "none",cex=1.2)
# 
# # get edge numbers for the species specific edges
# edgeNr <- setNames( match(match(spcs,spcTree$tip.label),spcTree$edge[ ,2]), spcs )
# 
# edgelabels(colSums(hSig),edge = branchEdgeNr,adj = c(0.5,1.4),cex=1.3,frame="none")


```

```{r pieTree, echo=FALSE,fig.height=3,fig.width=6}

# make the pamlMat have the same rows as hSig (call it PSmat)
# start with all NA
PSmat <- matrix(NA,ncol=ncol(pamlMat),nrow=nrow(hSig),dimnames = list(rownames(hSig),colnames(pamlMat)))
# fill in rows from pamlMat
grpIDsTestedHCOG <- rownames(pamlMat)[rownames(pamlMat) %in% rownames(PSmat)]
PSmat[grpIDsTestedHCOG, ] <- pamlMat[grpIDsTestedHCOG, ]

sapply(colnames(hSig), function(h){
  isTested <- !is.na(PSmat[,h]) # only include HCOGs that has been tested for positive selection (PS)
  isDE <- hSig[isTested,h]
  isPS <- PSmat[isTested,h]<0.05 
  pVal <- phyper(q = sum(isPS & isDE), # number DEGs with PS
                 m = sum(isPS), # total PS
                 n = sum(!isPS), # total not PS
                 k = sum(isDE), # total DEGs 
                 lower.tail=F) # test for overrepresentation of DEGs in PS

  return( c( nDE=sum(isDE), 
             nNotDE = sum(!isDE),
             nPSnotDE = sum(isPS & !isDE),
             nPSDE = sum(isPS & isDE),
             propPSDE = sum(isPS & isDE)/sum(isDE),
             propPSnotDE = sum(isPS & !isDE)/sum(!isDE),
             pVal = pVal) )  
}) -> pieStats

plotPosSelection <- function(){
  par(mar=c(1, 1, 1, 1) + 0.1)
  plot.phylo(spcTree,edge.width = 3,show.tip.label = F, x.lim=c(0,7),y.lim=c(0,7))
  
  for( i in seq_along(spcLabels)){
    tiplabels( bquote(italic(.(spcLabels[i]))), tip = match(spcTree$tip.label, names(spcLabels))[i] ,
               adj = 0,frame = "none",cex=1.2)  
  }
  
  # tiplabels( paste(" ",spcLabels), tip = match(spcTree$tip.label, names(spcLabels)) ,
  #            adj = 0,frame = "none",cex=1.2)
  
  edgelabels(edge = branchEdgeNr,
             pie = cbind(pieStats["propPSnotDE",],1-pieStats["propPSnotDE",]), 
             piecol=c("red","white"), cex=1.8)
  
  edgelabels(edge = branchEdgeNr,
             pie = cbind(pieStats["propPSDE",],1-pieStats["propPSDE",]), 
             piecol=c("purple","lightblue"), cex=1.1)
  
  
  edgelabels(paste0(pieStats["nPSnotDE",],"/",pieStats["nNotDE",]),
             edge = branchEdgeNr,adj = c(0.5,-2.4),frame="none",cex=0.8)
  
  edgelabels(paste0(pieStats["nPSDE",],"/",pieStats["nDE",]),
             edge = branchEdgeNr,adj = c(0.5,3.6),frame="none",cex=0.8)
  
  edgelabels(paste0("P = ",signif(pieStats["pVal",],2)),
             edge = branchEdgeNr,adj = c(0.5,4.9),frame="none",cex=0.8)
  par(mar=c(5, 4, 4, 2) + 0.1)
}

plotPosSelection() # once for the RMarkdown

svg(filename = "~/GRewd/pipeline/articleFigures/FigurePosSelection.svg",
    width = 6, height = 3)
plotPosSelection() # and once for the "official" figure
invisible(dev.off())
```






```{r dNdS, echo=FALSE, eval=F}
# dNdSall is a list with one entry for each orthoGrp that contains a list of three matrices dN, dS and dNdS

# Find pairwise distances between each species and rice..
# We want to know if the species specific DE genes have a higher dN/dS ratio than
# the non DE genes

# Is topology sorting needed? not at first
# What to do in case of multiple paralogs? select the first!
# Use de novo or reference core pooids? de novo..

spcs <- c("HoVu","BrDi","MeNu1","StLa","NaSt")
refSpc <- "Os_R"

lapply(c(dN="dN",dS="dS",dNdS="dNdS"),function(whichDist){
  sapply(setNames(dNdSall,sub("_.*$","",names(dNdSall))), function(allDist){
    m <- allDist[[whichDist]]
    m <- m+t(m)
    colnames(m) <- sub("\\|.*$","",colnames(m))
    idxSpcs <- match(spcs,colnames(m))
    idxRef <- match(refSpc,colnames(m))
    return(m[idxRef,idxSpcs])
  }) -> d
  d[d==-1] <- NA
  return(d)
}) -> Os_d

allDist <- dNdSall[[123]]
whichDist = "dN"
Os_d$dNdS[,1:5]

boxplot(t(Os_d$dNdS))

Os_d$dN[,1:5]
Os_d$dS[,1:5]

dNdS <- t(Os_d$dNdS[,match(rownames(spcSig),colnames(Os_d$dNdS))])
dNdS[1:5,]
spcSig[1:5,]

spc="HoVu"
x <- dNdS[spcSig[ ,spc],spc]
x2 <- dNdS[!spcSig[ ,spc],spc]

plot(density(x2,na.rm = T))
lines(density(x,na.rm = T),col="red")


plot(density(Os_dNdS[1,],na.rm = T))
for(i in 2:5) lines(density(Os_dNdS[i,],na.rm = T),col=i)

```

