---
title: "Tree with candidate LTI gene groups"
output: html_document
---


```{r loadLibraries, echo=FALSE, message=FALSE}
library(DESeq2)
library(gplots)
library(stringr)
library(ape)


source("~/GRewd/pipeline/R/orthoGrpTools.R")
```


```{r loadData, echo=FALSE}
orthoPath <- "/mnt/NOBACKUP/mariansc/share/orthos"


dNdSall <- readRDS("~/GRewd/pipeline/data/dNdSall.RDS") # from pamlResultsOverview.Rmd
pamlRes <- readRDS("~/GRewd/pipeline/data/pamlRes.RDS") # from pamlResultsOverview.Rmd 

hTestable <- readRDS("~/GRewd/pipeline/data/hTestable2.RDS") # from treeAnalysis6.Rmd

# convert pamlResult into pvalue matrix with only testable hypotheses
pamlMat <- matrix(NA_real_,nrow = nrow(hTestable),ncol = ncol(hTestable), 
                  dimnames = dimnames(hTestable))

for(h in colnames(hTestable)){
  testableGrpIDs <- rownames(hTestable)[hTestable[ ,h]]
  pamlMat[testableGrpIDs,h] <- pamlRes[[h]]$pVal[match(testableGrpIDs, pamlRes[[h]]$grpID)]
}

# load DEmat based based on superGenes
DEmat <- readRDS("~/GRewd/pipeline/data/DEmat.RDS")  # from: superGeneModel.Rmd

# load variance stabilized lolium (Falster) expression
loliumVSD <- read.table("~/GRewd/pipeline/indata/LoliumVSD/Fal_leaf_DE_VSD.txt",header=T,sep="\t",row.names = 1)


# load groups
goodGrps <- loadOrthoGrpsArray(file.path(orthoPath,"splitGroups/goodGroups.txt"))

```


### Combined peak and ramp

A gene is considered low temperature induced (LTI) if it is significantly differencially expressed (P-adjusted < 0.05 and abs(log fold change) > 1) either in peak or ramp.

```{r sigVennPlet, echo=FALSE}
# Get significant genes for peak and ramp (NA values equals not significant)
sigMatUpDown <- lapply(DEmat, with, { ifelse(is.na(pAdj), F, pAdj < 0.05 & abs(lfc) > 1) })
# Combine peak and ramp (significant if significant in either peak or ramp)
combSigMat <- sigMatUpDown$peak | sigMatUpDown$ramp

# plot Venn diagram
venn(apply(combSigMat,2,which))
title(main="significant peak or ramp (up or down)",sub="p.adj<0.05 & |lfc|>1")
```


```{r lolium, echo=FALSE, eval=FALSE}
# ### Lolium data
# 
# Data from the article [Vernalization Mediated Changes in the Lolium perenne Transcriptome](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0107365)
# 
# We used the data from leafs of the Falster variant.
#  




# convert lolium sequence ID's
loliumVSDseqIDs <- sub("comp([0-9]+)_c([0-9]+)_seq([0-9]+)","f\\1_c\\2s\\3",rownames(loliumVSD),perl=T)

# how many have ortholog in good group?
table(loliumGeneInGoodOrthoGroup=loliumVSDseqIDs %in% unlist(goodGrps[ ,"LoPeF"]))

# get grpIDs
loliumVSDgrpIDs <- seqIDtoGrpID(goodGrps,loliumVSDseqIDs,"LoPeF")

# How many sequences per group (paralogs)?
table(paralogCount=table(na.omit(loliumVSDgrpIDs)))

# get unique grps
loliumSigGrpIDs <- unique(na.omit(loliumVSDgrpIDs))

# How many of these groups are included in the superGeneMatrix?
table(loliumGrpIdInSuperGeneMatrix = loliumSigGrpIDs %in% rownames(combSigMat))


# # create a matrix with only
# loliumVSDmat <- as.matrix(loliumVSD)
# rownames(loliumVSDmat) <- loliumVSDseqIDs
# loliumVSDmat <- loliumVSDmat[rownames(loliumVSDmat) %in% unlist(goodGrps[ ,"LoPeF"]),]
# loliumVSDmatGrpIDs <- seqIDtoGrpID(goodGrps,rownames(loliumVSDmat),"LoPeF")

loliumSig <- rownames(combSigMat) %in% loliumSigGrpIDs


# plot Venn diagram
venn(apply(combSigMat & !loliumSig,2,which))
title(main="significant but not in lolium")

venn(apply(combSigMat & loliumSig,2,which))
title(main="significant also in lolium",
      sub=paste("Only in lolium =",sum(loliumSig & !apply(combSigMat,1,any))))


```


```{r calcSig, echo=FALSE}
# get species specific LTI genes
spcs <- colnames(combSigMat)
sapply( spcs, function(spc){
  # significant in spc and not significant in any other spc 
  combSigMat[ ,spc] & !apply(combSigMat[ ,spcs != spc],1,any)
}) -> spcSig


# Species included in each branch/hypothesis (excluding lolium)
hSpcs2 <- list(
  H4b = c("HoVu","BrDi"),
  #H4e = c("HoVu","BrDi","MeNu1"),
  H4d = c("HoVu","BrDi","MeNu1","StLa"),
  H4a = c("HoVu","BrDi","MeNu1","StLa","NaSt")
)

# Get branch specific LTI genes
sapply( hSpcs2, function(branchSpcs){
  # significant in all branch species and not significant in any other species 
  apply(combSigMat[ ,spcs %in% branchSpcs],1,all) & 
    !apply(combSigMat[ ,!(spcs %in% branchSpcs),drop=F],1,any)
}) -> hSig

```

```{r treePlot, echo=FALSE}
# spcTree <- read.tree(text = "((((((LoPe:1,HoVu:1)H4c:1,BrDi:2)H4b:1,MeNu1:3)H4e:1,StLa:4)H4d:1,NaSt:5)H4a:1,Os_R:6):1;")

# spcTree <- read.tree(text = "(((((HoVu:2,BrDi:2)H4b:1,MeNu1:3)H4e:1,StLa:4)H4d:1,NaSt:5)H4a:1,Os_R:6):1;")

# New species tree with collapse Stipa/Melica
spcTree <- read.tree(text = "((((HoVu:1,BrDi:1)H4b:1,MeNu1:2,StLa:2)H4d:1,NaSt:3)H4a:1,Os_R:4):0;")


# convert the 
spcLabels <- c( Os_R = "Oryza",
                NaSt = "Nardus",
                StLa = "Stipa",
                MeNu1 = "Melica",
                BrDi = "Brachypodium",
                HoVu = "Hordeum") #,LoPe = "Lolium")
                
plot.phylo(spcTree,edge.width = 3,show.tip.label = F, x.lim=c(0,10))
tiplabels( paste(" ",spcLabels), tip = match(spcTree$tip.label, names(spcLabels)) ,
           adj = 0,frame = "none",cex=1.2)

# get edge numbers for the species specific edges
edgeNr <- setNames( match(match(spcs,spcTree$tip.label),spcTree$edge[ ,2]), spcs )
# plot number of species specific LTI genes
edgelabels(colSums(spcSig),edge = edgeNr,adj = c(0.5,-0.4),cex=1.3,frame="none")

# get edge numbers for the branch specific edges
branchNodeNr <- match(colnames(hSig),spcTree$node.label) + length(spcTree$tip.label)
branchEdgeNr <- setNames( match(branchNodeNr,spcTree$edge[ ,2]), colnames(hSig) )

edgelabels(colSums(hSig),edge = branchEdgeNr,adj = c(0.5,1.4),cex=1.3,frame="none")


# # Exclusive in lolium
# loliumEdgeNr <- match(match("LoPe",spcTree$tip.label),spcTree$edge[ ,2])
# edgelabels(sum(loliumSig & !apply(combSigMat,1,any)) ,edge = loliumEdgeNr,
#            adj = c(0.5,-0.4),cex=1.3,frame="none", col="red")
# 
# # Exclusive in lolium and HoVu (H4c)
# sigH4c <- loliumSig & combSigMat[,"HoVu"] & !apply(combSigMat[,colnames(combSigMat) != "HoVu"],1,any)
# branchH4cEdgeNr <- match(match("H4c",spcTree$node.label) + length(spcTree$tip.label),
#                          spcTree$edge[ ,2])
# edgelabels(sum(sigH4c),edge = branchH4cEdgeNr,adj = c(0.5,1.4),cex=1.3,frame="none", col="red")
# 
# hSigPlus <- cbind(H4c=sigH4c,hSig)
hSigPlus <- cbind(hSig)

sapply(colnames(hSigPlus),function(h){
  grpIDsSigBranch <- rownames(hSigPlus)[hSigPlus[,h]]
  
  # how many are testable?
  #table(grpIDsSigBranch %in% rownames(hTestable))
  grpIDs <- grpIDsSigBranch[grpIDsSigBranch %in% rownames(hTestable)]
  
  nLTItested <- sum(!is.na(pamlMat[grpIDs,h]))
  nLTIpositive <- sum(na.omit(pamlMat[grpIDs,h]<0.05))
  
  #sum(na.omit(pamlMat[ ,h]<0.05))/length(na.omit(pamlMat[ ,h]))
  
  pVal <- phyper(q = nLTIpositive, # number LTI that tested significant
                 m = sum(na.omit(pamlMat[ ,h]<0.05)), # number tested that was significant
                 n = sum(na.omit(pamlMat[ ,h]>=0.05)), # number tested that was not significant
                 k = nLTItested, # number of LTI that have been tested
                 lower.tail=F)
  
#  paste0(round(100*nLTIpositive/nLTItested,digits = 2),"%(p=",round(pVal,2),")")
  paste0(nLTIpositive,"/",nLTItested,ifelse(pVal<0.05,"*",""))
}) -> branchPosSel

# edgelabels(branchPosSel,edge = c(branchH4cEdgeNr,branchEdgeNr),adj = c(0.5,2.8),frame="none")
edgelabels(branchPosSel,edge = branchEdgeNr,adj = c(0.5,2.8),frame="none")

```

```{r pieTree, echo=FALSE}

sapply(colnames(hSig), function(h){
  isTestable <- !is.na(pamlMat[,h]) & rownames(pamlMat) %in% rownames(hSig)

  pamlSig <- pamlMat[isTestable,h] < 0.05
  DESig <- hSig[match(rownames(pamlMat)[isTestable],rownames(hSig))]
  
  return( c( nDE=sum(DESig), 
             nTot=sum(isTestable), 
             propPamlDE = sum(pamlSig & DESig)/sum(DESig),
             propPamlNotDE = sum(pamlSig & !DESig)/sum(!DESig) ))  
}) -> pieStats


plot.phylo(spcTree,edge.width = 3,show.tip.label = F, x.lim=c(0,7))
tiplabels( paste(" ",spcLabels), tip = match(spcTree$tip.label, names(spcLabels)) ,
           adj = 0,frame = "none",cex=1.2)

edgelabels(edge = branchEdgeNr,
           pie = cbind(pieStats["propPamlNotDE",],1-pieStats["propPamlNotDE",]), 
           piecol=c("red","white"), cex=1.8)

edgelabels(edge = branchEdgeNr,
           pie = cbind(pieStats["propPamlDE",],1-pieStats["propPamlDE",]), 
           piecol=c("purple","lightblue"), cex=1.1)


```






```{r dNdS, echo=FALSE, eval=F}
# dNdSall is a list with one entry for each orthoGrp that contains a list of three matrices dN, dS and dNdS

# Find pairwise distances between each species and rice..
# We want to know if the species specific DE genes have a higher dN/dS ratio than
# the non DE genes

# Is topology sorting needed? not at first
# What to do in case of multiple paralogs? select the first!
# Use de novo or reference core pooids? de novo..

spcs <- c("HoVu","BrDi","MeNu1","StLa","NaSt")
refSpc <- "Os_R"

lapply(c(dN="dN",dS="dS",dNdS="dNdS"),function(whichDist){
  sapply(setNames(dNdSall,sub("_.*$","",names(dNdSall))), function(allDist){
    m <- allDist[[whichDist]]
    m <- m+t(m)
    colnames(m) <- sub("\\|.*$","",colnames(m))
    idxSpcs <- match(spcs,colnames(m))
    idxRef <- match(refSpc,colnames(m))
    return(m[idxRef,idxSpcs])
  }) -> d
  d[d==-1] <- NA
  return(d)
}) -> Os_d

allDist <- dNdSall[[123]]
whichDist = "dN"
Os_d$dNdS[,1:5]

boxplot(t(Os_d$dNdS))

Os_d$dN[,1:5]
Os_d$dS[,1:5]

dNdS <- t(Os_d$dNdS[,match(rownames(spcSig),colnames(Os_d$dNdS))])
dNdS[1:5,]
spcSig[1:5,]

spc="HoVu"
x <- dNdS[spcSig[ ,spc],spc]
x2 <- dNdS[!spcSig[ ,spc],spc]

plot(density(x2,na.rm = T))
lines(density(x,na.rm = T),col="red")


plot(density(Os_dNdS[1,],na.rm = T))
for(i in 2:5) lines(density(Os_dNdS[i,],na.rm = T),col=i)

```

